src/data/thunks.js

```diff
@@ -0,0 +1,105 @@
+import { logError } from '@edx/frontend-platform/logging';
+import {
+  fetchExamAttemptsData,
+  createExamAttempt,
+  stopAttempt,
+  continueAttempt,
+  submitAttempt,
+} from './api';
+import { isEmpty } from '../helpers';
+import { setIsLoading, setExamState, expireExamAttempt } from './slice';
+
+function updateAttemptAfter(courseId, sequenceId, promise = null, noLoading = false) {
+  return async (dispatch) => {
+    let data;
+    if (!noLoading) { dispatch(setIsLoading({ isLoading: true })); }
+    if (promise) {
+      data = await promise;
+      if (!data || !data.exam_attempt_id) {
+        if (!noLoading) { dispatch(setIsLoading({ isLoading: false })); }
+        return;
+      }
+    }
+
+    const attemptData = await fetchExamAttemptsData(courseId, sequenceId);
+    dispatch(setExamState({
+      exam: attemptData.exam,
+      activeAttempt: !isEmpty(attemptData.active_attempt) ? attemptData.active_attempt : null,
+    }));
+    if (!noLoading) { dispatch(setIsLoading({ isLoading: false })); }
+  };
+}
+
+export function getExamAttemptsData(courseId, sequenceId) {
+  return updateAttemptAfter(courseId, sequenceId);
+}
+
+export function startExam() {
+  return async (dispatch, getState) => {
+    const { exam } = getState().examState;
+    if (!exam.id) {
+      logError('Failed to start exam. No exam id.');
```

Comment:
What's the user experience when this or other errors occur? We don't have to address that here, but we should have a story for it.

Reply:
Created another ticket for that, suggesting to deliver it with proctored exams